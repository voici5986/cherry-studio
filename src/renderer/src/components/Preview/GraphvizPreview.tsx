import { AsyncInitializer } from '@renderer/utils/asyncInitializer'
import React, { memo, useCallback } from 'react'
import styled from 'styled-components'

import { useDebouncedRender } from './hooks/useDebouncedRender'
import ImagePreviewLayout from './ImagePreviewLayout'
import { BasicPreviewHandles, BasicPreviewProps } from './types'
import { renderSvgInShadowHost } from './utils'

// 管理 viz 实例
const vizInitializer = new AsyncInitializer(async () => {
  const module = await import('@viz-js/viz')
  return await module.instance()
})

/** 预览 Graphviz 图表
 * 使用 usePreviewRenderer hook 大幅简化组件逻辑
 */
const GraphvizPreview = ({
  children,
  enableToolbar = false,
  ref
}: BasicPreviewProps & { ref?: React.RefObject<BasicPreviewHandles | null> }) => {
  // 定义渲染函数
  const renderGraphviz = useCallback(async (content: string, container: HTMLDivElement) => {
    const viz = await vizInitializer.get()
    const svg = viz.renderString(content, { format: 'svg' })
    renderSvgInShadowHost(svg, container)
  }, [])

  // 使用预览渲染器 hook
  const { containerRef, error, isLoading } = useDebouncedRender(children, renderGraphviz, {
    debounceDelay: 300
  })

  return (
    <ImagePreviewLayout
      loading={isLoading}
      error={error}
      enableToolbar={enableToolbar}
      ref={ref}
      imageRef={containerRef}
      source="graphviz">
      <StyledGraphviz ref={containerRef} className="graphviz special-preview" />
    </ImagePreviewLayout>
  )
}

const StyledGraphviz = styled.div`
  overflow: auto;
  position: relative;
  width: 100%;
  height: 100%;
`

export default memo(GraphvizPreview)
