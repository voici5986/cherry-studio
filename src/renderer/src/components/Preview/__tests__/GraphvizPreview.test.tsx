import GraphvizPreview from '@renderer/components/Preview/GraphvizPreview'
import { render, screen } from '@testing-library/react'
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest'

// Use vi.hoisted to manage mocks
const mocks = vi.hoisted(() => ({
  vizInstance: {
    renderSVGElement: vi.fn()
  },
  vizInitializer: {
    get: vi.fn()
  },
  ImagePreviewLayout: vi.fn(({ children, loading, error, enableToolbar, source }) => (
    <div data-testid="image-preview-layout" data-source={source}>
      {enableToolbar && <div data-testid="toolbar">Toolbar</div>}
      {loading && <div data-testid="loading">Loading...</div>}
      {error && <div data-testid="error">{error}</div>}
      <div data-testid="preview-content">{children}</div>
    </div>
  )),
  useDebouncedRender: vi.fn()
}))

vi.mock('@renderer/components/Preview/ImagePreviewLayout', () => ({
  default: mocks.ImagePreviewLayout
}))

vi.mock('@renderer/utils/asyncInitializer', () => ({
  AsyncInitializer: class {
    constructor() {
      return mocks.vizInitializer
    }
  }
}))

vi.mock('@renderer/components/Preview/hooks/useDebouncedRender', () => ({
  useDebouncedRender: mocks.useDebouncedRender
}))

describe('GraphvizPreview', () => {
  const dotCode = 'digraph { a -> b }'
  const mockContainerRef = { current: document.createElement('div') }

  // Helper function to create mock useDebouncedRender return value
  const createMockHookReturn = (overrides = {}) => ({
    containerRef: mockContainerRef,
    error: null,
    isLoading: false,
    triggerRender: vi.fn(),
    cancelRender: vi.fn(),
    clearError: vi.fn(),
    setLoading: vi.fn(),
    ...overrides
  })

  beforeEach(() => {
    // Setup default successful state
    mocks.useDebouncedRender.mockReturnValue(createMockHookReturn())
  })

  afterEach(() => {
    vi.clearAllMocks()
  })

  describe('basic rendering', () => {
    it('should match snapshot', () => {
      const { container } = render(<GraphvizPreview enableToolbar>{dotCode}</GraphvizPreview>)
      expect(container).toMatchSnapshot()
    })

    it('should handle valid dot code', () => {
      render(<GraphvizPreview>{dotCode}</GraphvizPreview>)

      // Component should render without throwing
      expect(screen.getByTestId('image-preview-layout')).toBeInTheDocument()
      expect(mocks.useDebouncedRender).toHaveBeenCalledWith(
        dotCode,
        expect.any(Function),
        expect.objectContaining({ debounceDelay: 300 })
      )
    })

    it('should handle empty content', () => {
      render(<GraphvizPreview>{''}</GraphvizPreview>)

      // Component should render without throwing
      expect(screen.getByTestId('image-preview-layout')).toBeInTheDocument()
      expect(mocks.useDebouncedRender).toHaveBeenCalledWith('', expect.any(Function), expect.any(Object))
    })
  })

  describe('loading state', () => {
    it('should show loading indicator when rendering', () => {
      mocks.useDebouncedRender.mockReturnValue(createMockHookReturn({ isLoading: true }))

      render(<GraphvizPreview>{dotCode}</GraphvizPreview>)

      expect(screen.getByTestId('loading')).toBeInTheDocument()
    })

    it('should not show loading indicator when not rendering', () => {
      mocks.useDebouncedRender.mockReturnValue(createMockHookReturn({ isLoading: false }))

      render(<GraphvizPreview>{dotCode}</GraphvizPreview>)

      expect(screen.queryByTestId('loading')).not.toBeInTheDocument()
    })
  })

  describe('error handling', () => {
    it('should show error message when rendering fails', () => {
      const errorMessage = 'Invalid dot syntax'
      mocks.useDebouncedRender.mockReturnValue(createMockHookReturn({ error: errorMessage }))

      render(<GraphvizPreview>{dotCode}</GraphvizPreview>)

      const errorElement = screen.getByTestId('error')
      expect(errorElement).toBeInTheDocument()
      expect(errorElement).toHaveTextContent(errorMessage)
    })

    it('should not show error when rendering is successful', () => {
      mocks.useDebouncedRender.mockReturnValue(createMockHookReturn({ error: null }))

      render(<GraphvizPreview>{dotCode}</GraphvizPreview>)

      expect(screen.queryByTestId('error')).not.toBeInTheDocument()
    })
  })

  describe('ref forwarding', () => {
    it('should forward ref to ImagePreviewLayout', () => {
      const ref = { current: null }
      render(<GraphvizPreview ref={ref}>{dotCode}</GraphvizPreview>)

      // The ref should be passed to ImagePreviewLayout
      expect(mocks.ImagePreviewLayout).toHaveBeenCalledWith(expect.objectContaining({ ref }), undefined)
    })
  })
})
