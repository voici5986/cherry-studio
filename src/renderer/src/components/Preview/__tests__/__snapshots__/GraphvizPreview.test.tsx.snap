// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GraphvizPreview > basic rendering > should match snapshot 1`] = `
.c0 {
  overflow: auto;
  position: relative;
  width: 100%;
  height: 100%;
}

<div>
  <div
    data-source="graphviz"
    data-testid="image-preview-layout"
  >
    <div
      data-testid="toolbar"
    >
      Toolbar
    </div>
    <div
      data-testid="preview-content"
    >
      <div
        class="c0 graphviz special-preview"
      />
    </div>
  </div>
</div>
`;
